src/type-extensions/proxy.ts

```diff
@@ -0,0 +1,164 @@
+import { BaseDecorationOptions, Decoration } from '../decoration'
+import { LuaReturn, LuaState, LuaType } from '../types'
+import MultiReturn from '../multireturn'
+import Thread from '../thread'
+import TypeExtension from '../type-extension'
+
+export interface ProxyDecorationOptions extends BaseDecorationOptions {
+    disableProxy?: boolean
+}
+
+export function decorateProxy(target: any, options?: ProxyDecorationOptions): Decoration<any, ProxyDecorationOptions> {
+    return new Decoration<any, ProxyDecorationOptions>(target, options || {})
+}
+
+class ProxyTypeExtension extends TypeExtension<any, ProxyDecorationOptions> {
+    private readonly gcPointer: number
+
+    public constructor(thread: Thread) {
+        super(thread, 'js_proxy')
+
+        this.gcPointer = thread.cmodule.module.addFunction((functionStateAddress: LuaState) => {
+            // Throws a lua error which does a jump if it does not match.
+            const userDataPointer = thread.cmodule.luaL_checkudata(functionStateAddress, 1, this.name)
+            const referencePointer = thread.cmodule.module.getValue(userDataPointer, '*')
+            thread.cmodule.unref(referencePointer)
+
+            return LuaReturn.Ok
+        }, 'ii')
+
+        if (thread.cmodule.luaL_newmetatable(thread.address, this.name)) {
+            const metatableIndex = thread.cmodule.lua_gettop(thread.address)
+
+            // Mark it as uneditable
+            thread.cmodule.lua_pushstring(thread.address, 'protected metatable')
+            thread.cmodule.lua_setfield(thread.address, metatableIndex, '__metatable')
+
+            // Add the gc function
+            thread.cmodule.lua_pushcclosure(thread.address, this.gcPointer, 0)
+            thread.cmodule.lua_setfield(thread.address, metatableIndex, '__gc')
+
+            thread.pushValue((self: any, key: unknown) => {
+                switch (typeof key) {
+                    case 'number':
+                        // Map from Lua's 1 based indexing to JS's 0.
+                        // This is especially important here because ipairs just calls
+                        // __index with 1, 2, 3, 4 etc until there's a null.
+                        key = key - 1
+                    // Fallthrough
+                    case 'string':
+                        break
+                    default:
+                        throw new Error('Only strings or numbers can index js objects')
+                }
+
+                const value = self[key as string | number]
+                if (typeof value === 'function') {
```

Comment:
A thing I noticed is that classes also have the type of function, but they are not callable like one. Maybe we can check for that too. Classes have the property `prototype.constructor` equal to itself, functions not.

Reply:
Ahh I see, thanks, will patch

Reply:
Now done. Though I did a test and if you bind to a constructor it still works fine surprisingly

Reply:
The solutions doesn't seems to work because some functions like `fs.readFileSync` have the constructor pointing to itself too :(

Reply:
`(value?.prototype?.constructor !== value || !value.toString().startsWith('class '))` maybe?

Reply:
Just realized the function type extension have the same problem

Reply:
Do you have a demo that can reproduce the issue?

I just tried this and it worked
```js
    const factory = new LuaFactory()
    const engine = await factory.createEngine({ injectObjects: true })
    engine.global.set('fs', fs)

    const thread = engine.global.newThread()

    thread.loadString(`
        print(fs.readFileSync('./test.js'))
    `)

    await thread.run()
```

Reply:
The `readFileAsync` was not a good example, sorry. But it will be a problem when the `bind(self)` is needed, I could reproduce that with the demoon example, the `http.createServer.listen` have the constructor binded to itself but should be called as a function with the `this` bound.

Reply:
Weirdly I added that bit of code to the function type and 12 tests broke for no seemingly related reason.

Do you want to try and PR onto my PR or otherwise try and fix this? I'm not sure and you seem pretty confident about it

Reply:
Opened #30